cpp/dolfinx/common/subsystem.cpp

```diff
@@ -64,18 +63,26 @@ void subsystem::init_logging(int argc, char* argv[])
   loguru::init(argc, argv_copy.data(), options);
 }
 //-----------------------------------------------------------------------------
-void subsystem::init_petsc()
+void subsystem::init_petsc(MPI_Comm comm)
 {
   int argc = 0;
   char** argv = nullptr;
-  init_petsc(argc, argv);
+  init_petsc(argc, argv, comm);
 }
 //-----------------------------------------------------------------------------
-void subsystem::init_petsc(int argc, char* argv[])
+void subsystem::init_petsc(int argc, char* argv[], MPI_Comm comm)
 {
   if (argc > 1)
     LOG(INFO) << "Initializing PETSc with given command-line arguments.";
 
+  // Ensure MPI has been initialised, otberwise setting PETSC_COMM_WORLD can 
+  // call MPI not initialised errors.
+  init_mpi(argc, argv);
+  // Ensure the PETSC_COMM_WORLD is not MPI_COMM_WORLD if the user has specified 
+  // a communicator.
+  //
+  PETSC_COMM_WORLD = comm;
```

Comment:
Is this safe? Looks wrong to assign to `PETSC_COMM_WORLD`.

Reply:
Yes, it looks weird doesn't it, but it's the "PETSc" way, i.e. https://petsc.org/main/docs/manualpages/Sys/PETSC_COMM_WORLD.html